File: protocol/impls/ts/src/builders.ts
// Test/generator construction helpers — build carrier payloads + abstract txs for the fold.
import type { Bytes } from "./bytes.ts";
import { concat, u8, u32le, u64le, leBytes } from "./bytes.ts";
import { OP, PREFIX0, PREFIX1, PREFIX2 } from "./constants.ts";
import type { FoldInput, FoldCarrier, FoldOutput, FoldTx } from "./fold.ts";

const enc = new TextEncoder();
export const str = (s: string): Bytes => enc.encode(s);

// generator identity scheme (conformance §5): h160 = byte(i) ‖ 18 zero bytes ‖ byte(i)
export function genId(i: number): Bytes {
  const b = new Uint8Array(20);
  b[0] = i & 0xff;
  b[19] = i & 0xff;
  return b;
}
export const genType = (i: number): number => (i % 4 === 3 ? 1 : 0); // P2SH if i%4==3 else P2PKH

export function input(id: Bytes | null, type = 0, sighashAll = true): FoldInput {
  return { id, type, sighashAll };
}
export function output(type: number, hash: Bytes, value: bigint): FoldOutput {
  return { type, hash, value };
}
export function tx(inputs: FoldInput[], carriers: FoldCarrier[], outputs: FoldOutput[] = []): FoldTx {
  return { inputs, carriers, outputs };
}

function frame(op: number, body: Bytes): Bytes {
  return concat(u8(PREFIX0), u8(PREFIX1), u8(PREFIX2), u8(op), body);
}
export function carrier(op: number, body: Bytes, value = 0n, vout = 0): FoldCarrier {
  return { payload: frame(op, body), value, vout };
}
export function rawCarrier(payload: Bytes, value = 0n, vout = 0): FoldCarrier {
  return { payload, value, vout };
}

// ─── per-op body builders (names-only 0x01–0x0C) ──────────────────────────────────────────────
export const commit = (commitment: Bytes, vout = 0): FoldCarrier =>
  carrier(OP.COMMIT, commitment, 0n, vout);
export const claim = (salt: Bytes, name: string, burn: bigint, vout = 0): FoldCarrier =>
  carrier(OP.CLAIM, concat(salt, str(name)), burn, vout);
export const renewAll = (burn: bigint, vout = 0): FoldCarrier =>
  carrier(OP.RENEW, new Uint8Array(0), burn, vout);
export const renewAllSafe = (anchor: bigint, burn: bigint, vout = 0): FoldCarrier =>
  carrier(OP.RENEW, leBytes(anchor, 5), burn, vout);
export const renewSel = (anchor: bigint, flags: Bytes, burn: bigint, vout = 0): FoldCarrier =>
  carrier(OP.RENEW, concat(leBytes(anchor, 5), flags), burn, vout);
export const transferAll = (target: Bytes, vout = 0): FoldCarrier =>
  carrier(OP.TRANSFER, target, 0n, vout);
export const transferSel = (target: Bytes, anchor: bigint, flags: Bytes, vout = 0): FoldCarrier =>
  carrier(OP.TRANSFER, concat(target, leBytes(anchor, 5), flags), 0n, vout);
export const sell = (price: bigint, window: bigint, name: string, vout = 0): FoldCarrier =>
  carrier(OP.SELL, concat(u64le(price), u32le(window), str(name)), 0n, vout);
export const reserve = (name: string, value: bigint, vout = 0): FoldCarrier =>
  carrier(OP.RESERVE, str(name), value, vout);
export const settle = (name: string, vout = 0): FoldCarrier =>
  carrier(OP.SETTLE, str(name), 0n, vout);
export const release = (anchor: bigint, flags: Bytes, vout = 0): FoldCarrier =>
  carrier(OP.RELEASE, concat(leBytes(anchor, 5), flags), 0n, vout);
export const sellTo = (price: bigint, buyer: Bytes, name: string, vout = 0): FoldCarrier =>
  carrier(OP.SELL_TO, concat(u64le(price), buyer, str(name)), 0n, vout);
export const pay = (name: string, vout = 0): FoldCarrier =>
  carrier(OP.PAY, str(name), 0n, vout);
export const asMarker = (index: number, vout = 0): FoldCarrier =>
  carrier(OP.AS, u8(index), 0n, vout);
export const trade = (idxA: number, idxB: number, nameA: string, nameB: string, vout = 0): FoldCarrier =>
  carrier(OP.TRADE, concat(u8(idxA), u8(idxB), str(nameA), str(",") , str(nameB)), 0n, vout);

// commitment = SHA256(salt ‖ name ‖ author_h160) — exported for test construction
import { sha256 } from "./sha256.ts";
export const commitmentOf = (salt: Bytes, name: string, author: Bytes): Bytes =>
  sha256(concat(salt, str(name), author));

File: protocol/impls/csharp/Builders.cs
using System;
using System.Buffers.Binary;
using System.Collections.Generic;

namespace Pepenet;

/// <summary>Convenience builders for hand-authored vectors (selftest / scenario).</summary>
public static class B
{
    // generator identity convention (SPEC-conformance §5): h160 = byte(i) ‖ 18 zero ‖ byte(i)
    public static byte[] Id(int i) { var h = new byte[20]; h[0] = (byte)i; h[19] = (byte)i; return h; }
    public static byte IdType(int i) => (i % 4 == 3) ? K.TYPE_P2SH : K.TYPE_P2PKH;

    public static Input In(int id, bool sah = true) => new() { H160 = Id(id), Type = IdType(id), SighashAll = sah };
    public static Out Spend(int id, ulong val) => Out.Spend(Id(id), IdType(id), val);
    public static Out Spend(byte[] h160, byte type, ulong val) => Out.Spend(h160, type, val);

    public static byte[] Name(string s) => System.Text.Encoding.ASCII.GetBytes(s);

    public static byte[] Payload(byte op, byte[] body)
    {
        byte[] p = new byte[4 + body.Length];
        p[0] = K.PFX0; p[1] = K.PFX1; p[2] = K.PFX2; p[3] = op;
        Buffer.BlockCopy(body, 0, p, 4, body.Length);
        return p;
    }

    public static byte[] U32(uint v) { byte[] b = new byte[4]; BinaryPrimitives.WriteUInt32LittleEndian(b, v); return b; }
    public static byte[] U64(ulong v) { byte[] b = new byte[8]; BinaryPrimitives.WriteUInt64LittleEndian(b, v); return b; }
    public static byte[] U40(long v) { byte[] b = new byte[5]; for (int i = 0; i < 5; i++) b[i] = (byte)(v >> (8 * i)); return b; }

    public static byte[] Concat(params byte[][] parts)
    {
        int n = 0; foreach (var p in parts) n += p.Length;
        byte[] r = new byte[n]; int o = 0;
        foreach (var p in parts) { Buffer.BlockCopy(p, 0, r, o, p.Length); o += p.Length; }
        return r;
    }

    public static byte[] Salt(byte b) { var s = new byte[32]; for (int i = 0; i < 32; i++) s[i] = b; return s; }

    public static byte[] Commitment(byte[] salt, byte[] name, byte[] author)
        => Hashing.Sha256(Concat(salt, name, author));

    // ---- payload builders ----
    public static byte[] Commit(byte[] commitment) => Payload(K.OP_COMMIT, commitment);
    public static byte[] Claim(byte[] salt, string name) => Payload(K.OP_CLAIM, Concat(salt, Name(name)));
    public static byte[] RenewAll() => Payload(K.OP_RENEW, Array.Empty<byte>());
    public static byte[] RenewAllSafe(long anchor) => Payload(K.OP_RENEW, U40(anchor));
    public static byte[] RenewSel(long anchor, byte[] flags) => Payload(K.OP_RENEW, Concat(U40(anchor), flags));
    public static byte[] TransferAll(byte[] target) => Payload(K.OP_TRANSFER, target);
    public static byte[] TransferSel(byte[] target, long anchor, byte[] flags) => Payload(K.OP_TRANSFER, Concat(target, U40(anchor), flags));
    public static byte[] Release(long anchor, byte[] flags) => Payload(K.OP_RELEASE, Concat(U40(anchor), flags));
    public static byte[] Sell(ulong price, uint window, string name) => Payload(K.OP_SELL, Concat(U64(price), U32(window), Name(name)));
    public static byte[] Reserve(string name) => Payload(K.OP_RESERVE, Name(name));
    public static byte[] Settle(string name) => Payload(K.OP_SETTLE, Name(name));
    public static byte[] SellTo(ulong price, byte[] buyer, string name) => Payload(K.OP_SELL_TO, Concat(U64(price), buyer, Name(name)));
    public static byte[] Pay(string name) => Payload(K.OP_PAY, Name(name));
    public static byte[] As(byte index) => Payload(K.OP_AS, new[] { index });
    public static byte[] Trade(byte idxA, byte idxB, string nameA, string nameB)
        => Payload(K.OP_TRADE, Concat(new[] { idxA, idxB }, Name(nameA), new byte[] { 0x2C }, Name(nameB)));
}
